python/examples/core/3-managing-gradients.py

```diff
@@ -0,0 +1,289 @@
+# set of values and their corresponding gradients.
+#
+# Let's compute the square of the radial spectrum, :math:`h(r) = \rho^2(r)`, and the
+# corresponding gradients with respect to atomic positions. The chain rules tells us
+# that the gradient should be
+#
+# .. math::
+#
+#     \nabla h(r) = 2\ \rho(r) * \nabla \rho(r)
+#
+# Since the calculation can happen block by block, let's define a function to compute a
+# new :math:`h(r)` block:
+
+
+def compute_square(block: TensorBlock) -> TensorBlock:
+    # compute the new values
+    new_values = block.values**2
+
+    # store the new values in a block
+    new_block = TensorBlock(
+        values=new_values,
+        samples=block.samples,
+        components=block.components,
+        properties=block.properties,
+    )
+
+    # compute the new gradient
+    gradient = block.gradient("positions")
+
+    # `block.values[gradient.samples["sample"]]` gives us an array with a shape
+    # compatible with `gradient.values`; using the right row in the values to compute
+    # the a given row of the gradients. ``None`` creates an additional dimension to
+    # match the components of the gradients.
+    broadcasted_values = block.values[gradient.samples["sample"], None, :]
+    new_gradient_values = 2.0 * broadcasted_values * gradient.values
+
+    new_gradient = TensorBlock(
+        values=new_gradient_values,
+        samples=gradient.samples,
+        components=gradient.components,
+        properties=gradient.properties,
+    )
+
+    # store the gradient in the new block
+    new_block.add_gradient("positions", new_gradient)
+
+    return new_block
+
+
+# %%
+#
+# One issue when applying the equation above blindly is that ``block.values`` (i.e.
+# :math:`\rho(r)`) and ``gradient.values`` (i.e. :math:`\nabla \rho(r)`) have different
+# shape. Fortunately, we already know how to match them: ``gradient.samples["sample"]``
+# contains the indices of ``block.values`` matching each row of ``gradient.values``.
+
+gradient = radial_spectrum.block(2).gradient("positions")
+print(gradient.samples["sample"])
+
+# %%
+#
+# We can now apply this function on all the blocks, and reconstruct a new
+# :py:class:`TensorMap`:
+
+blocks = [compute_square(block) for block in radial_spectrum.blocks()]
+squared = TensorMap(keys=radial_spectrum.keys, blocks=blocks)
+
+# %%
+#
+# ``squares`` has the same shape and sparsity pattern as ``radial_spectrum``, but
+# contains different values:
+
+print(squared)
+
+# %%
+#
+rs_block = radial_spectrum.block(2)
+squared_block = squared.block(2)
+
+print("radial_spectrum block:", rs_block)
+print("square block:", squared_block)
+
+
+# %%
+#
+
+print("radial_spectrum values:", rs_block.values)
+print("square values:", squared_block.values)
+
+# %%
+#
+
+print("radial_spectrum gradient:", rs_block.gradient("positions").values[:, 0])
+print("square gradient:", squared_block.gradient("positions").values[:, 0])
```

Comment:
Very nice example with the function. We should mention though that we have metatensor operations that take gradients propperly into account and users don't have to write basic operations by themselves!

Reply:
good point!